[MVP][WIP] Less like pager

zhiburt opened the pull request:
Run it as `tabless`.

#### example

```nu
ls | tabless
```

webbedspace commented:
I don't like the pun name, for these reasons:
* Too difficult to "get" if you don't already know what the two components are (otherwise, looks like a typo of `tables`).
* Personally don't like the name of `table` due to the word "table" describing too many things in Nushell (#6772) and this doubles down on it.
* `less` itself is starting to fall out of favour in place of third-party alternatives like `bat`, so the name's resonance (in terms of describing the concept of pagers itself) is fading.
* Pun names are typically the domain of third-party programs trying to have memorable branding (e.g. "Ag, the Silver Searcher") and not internal commands of a utility. Typical Nushell command naming is usually a more direct verb instead of a POSIX name (`find` instead of `grep`, `fetch` instead of `curl` etc.) with exceptions only given to directory navigation (`mkdir` etc.)

My proposal is renaming it to something like `pager` or `scroll`.

fdncred commented:
@zhiburt This is my torture test for all things related to tables. :)
opening this file reveals some interesting artifacts.
[em.csv](https://github.com/nushell/nushell/files/9929401/em.csv)

zhiburt commented:
I did a few improvements here and there and I think it's more or less in a good state now.

A few questions;

1. What should it return to the pipe? (`PipelineState`)
    I think nothing?

2. I am thinking to add `PG_UP`, `PG_DOWN` keys to scroll pages. But is there more useful commands?

> opening this file reveals some interesting artifacts.

I've run through some part of it, and seems OK.
Could you show it? (I am sure there will be ones)

> I don't like the pun name, for these reasons:

Good points; @webbedspace

fdncred commented:
@zhiburt I think PGUP and PGDOWN are good and also letting `esc`, `ctrl-c` along with `ctrl-d` to exit it.

Notice the artifacts of lines on the right side and sometimes text not erasing as well as the columns jumping around. not sure if there's anything you can do about these things but it's what i was talking about.
![scroll](https://user-images.githubusercontent.com/343840/199793719-9e13ed2c-b496-425b-ace2-6ed2da0cb9c5.gif)

zhiburt commented:
> Notice the artifacts of lines on the right side and sometimes text not erasing as well as the columns jumping around. not sure if there's anything you can do about these things but it's what i was talking about.

Interesting....

Could you check whether it was fixed?

fdncred commented:
> Could you check whether it was fixed?

Not fixed. It still exhibits the same artifacts. I tried in Windows Terminal and in Tabby.

zhiburt commented:
> > Could you check whether it was fixed?
> 
> Not fixed. It still exhibits the same artifacts. I tried in Windows Terminal and in Tabby.

that's interesting.

__________________________

I've spend some time to do a recursive logic to get inside inner values.
(yet not completed).

But wanted to know whether it's worth to continue.

![image](https://user-images.githubusercontent.com/20165848/200090414-d1907b9e-b9c3-461e-8c63-0f7ee62704d3.png)

The idea is if the value is actually a record/list we can inspect it further in a pop up instead of having `{ record .. }`\`[ list ... ]` message.

PS: Personally I kind of feel it's more logic than a simple pager like `less`/`more` shall have. So it may be better splitted as a completely different binary.

fdncred commented:
Oh wow! Ya, that's totally cool. In think being able to inspect things is a good add.

zhiburt commented:
I hope I made it work.

![image](https://user-images.githubusercontent.com/20165848/200143703-f18e7450-7d4f-4fd5-9d6b-049e0189f33a.png)

> Press `i` to get in a 'cursor' mode
> Use 'Arrow keys' to select a cell you want
> Press enter, and you must get in inner view.
> To exit it press `ESC` 

PS: Haven't yet looked at your artifact issue.

fdncred commented:
Nice, this is cool. So, if the list is a table it shows like `ls -al | scroll` shows but if it has nested collections, it allows you to go into them with `i`. Is that how it works?

zhiburt commented:
Not sure if you ran it;
Here's a gif.

I've changed a an inner view.

![image](https://user-images.githubusercontent.com/20165848/200161567-f7dac016-8230-4080-acec-8ebc8c1a2ce6.png)

zhiburt commented:
> I did $nu.scope.commands | scroll hit i and went down to ansi gradient then scrolled over to signature and hit enter and got this screen. Is there a way to scroll over to the right where it says custom_...? I couldn't get it to scroll.

So you need to press `ESC` once to change mode back to 'view' then you'll must be able to do so.

> It would be cool to hit / and be able to search in whichever column you're in.

Added some support of it, though not 'column based'.

1. Press `/` to enter the search hit `Enter` next;
2. You can iterate over results by pressing `n` 

> Another suggestion, It would be nice to have something other than a blinking pipe to show your current selection, like highlight the entire row in reverse video or choose some background color for the whole row.

Arguable thing, but surely being able to make it more customarily could be great. (like whether you you use cursor or selection etc.)

PS: I think you search in column is good idea, I also can picture different kind of sorts could be beneficial too (like search by the whole structure so if we found something somewhere deep in the value tree we would be able to get in there. right from search)

fdncred commented:
nice - looks good! it's really fun to play with.

zhiburt commented:
Added a customization options.

![image](https://user-images.githubusercontent.com/20165848/200824109-7348af9a-ea89-4e1b-8569-3f2211c9759d.png)

https://github.com/nushell/nushell/blob/8f43e3f01a66a815ff7c1ae0615083c427326bfe/crates/nu-utils/src/sample_config/default_config.nu#L283-L293

PS: Somehow when colors set to `{bg: '..', fg: '..'}` it doesn't work cause it considers it being a string. I wonder if I am doing something wrong or there's some limit (2 deep max?) for config.

So this works `let-env config = { scroll_config: { highlight: 'red' } }`
But this doesn't `let-env config = { scroll_config: { highlight: { fg: 'red'} } }`

fdncred commented:
> Now I see it's just that string values are not supported in fg/bg right?

you are right that named colors are not supported with this syntax, only hex colors. although, we'd accept a PR for that if anyone watching is interested in adding that. it's been asked a few times.

fdncred commented:
I can't get this to work to test it
```
   scroll_config {
     highlight: { bg: '#ff0000', fg: '#ffffff' }
     selected_column: 'blue'
     selected_row: 'yellow'
     selected_cell: 'green'
     status_bar: 'blue'
     command_bar: 'blue'
     split_line: 'green'
     cursor: false
   }
```
I'm not sure what's going on right now

zhiburt commented:
>  I'm not sure what's going on right now

To be honest for some reason if I update my config I also had no changes.....
I've played with this via.

```nu
let-env config = { table_mode: basic, scroll_config: { highlight: 'yellow', selected_column: 'blue', selected_row: 'yellow', selected_cell: { fg: '#333333', bg: '#777777' }, cursor: false } }
```

zhiburt commented:
> you are right that named colors are not supported with this syntax, only hex colors. although, we'd accept a PR for that if anyone watching is interested in adding that. it's been asked a few times.

Must be addressed

fdncred commented:
oh wow, i didn't realize this actually works. I must of missed a PR
```
$"(ansi { fg: yellow bg: red attr: b})Hello(ansi reset)"
```
![image](https://user-images.githubusercontent.com/343840/200873425-fc19175b-c205-47f2-aced-2f5541e89a49.png)

fdncred commented:
not quite the right colors but i'm starting to love this color customization feature.
```
     # A 'scroll' utility config
    scroll_config: {
        highlight: { bg: yellow, fg: black }
        status_bar: { bg: '#C4C9C6', fg: '#1D1F21' }
        command_bar: { fg: '#C4C9C6' }
        split_line: white
        cursor: false
        selected_column: light_green_reverse
        selected_row: { fg: white bg: dark_gray }
        selected_cell: { fg: black bg: light_green }
    }
```
![scroll2](https://user-images.githubusercontent.com/343840/200877200-2c24d5b7-0891-48b0-80e1-9799956530ed.gif)

remmycat commented:
Had another thought on this, in the hope that this doesn't explode the scope of the PR too much (if it does please ignore 😅)

I would find it extremely cool, if only for the teaching aspect, if the TUI would always show the command to get to the current view.

e.g.
`table | get columname.10`
`table | find 'searchterm'`
`table | find 'searchterm' | get columname.5 | find 'other'`

This would be immensely powerful if people can then even edit this to do use more complex filters, but I think that's definitely not trivial to implement.

zhiburt commented:
> With the added functionality for filtering columns, and "opening" cells, I was wondering if `scroll` is still an accurate name?
> 
> How about something more general, like `explore`? That would also give it much more flexibility when it comes to adding more features over time, without having to justify why this is part of the `scroll` command.

That's exactly what I am up to as well.
I mean maybe it's worth to keep 2 separate programs 1 with only scrolling mode and the second more sophisticated.
I don't know.

I like the name.



> Had another thought on this, in the hope that this doesn't explode the scope of the PR too much (if it does please ignore sweat_smile)
> 
> I would find it extremely cool, if only for the teaching aspect, if the TUI would always show the command to get to the current view.
> 
> e.g. `table | get columname.10` `table | find 'searchterm'` `table | find 'searchterm' | get columname.5 | find 'other'`
> 
> This would be immensely powerful if people can then even edit this to use more complex filters and such. I guess that could be done by having a command to exit the view and "apply" (append) the generated filter string to your last command in history, or a temp value that was saved. (No need to reimplement the entire shell inside the tui.)

That's seems like achievable.
The only thing is I am not aware if `nu`shell actually keeps the whole command through it's pipelines?

And yes there's a question about the interface of it? I mean a shortcut?

You left me with a good though though,
look at this.

![download (6)](https://user-images.githubusercontent.com/20165848/201206433-47b751bd-383b-4d9b-a441-b0e7955e3f8d.gif)

The only thing I am not sure whether we better hidden this behavior by a flag or shall be a default and a reverse flag shall be introduced?

fdncred commented:
oh nice! i like how you can output a column.

dandavison commented:
This branch looks extremely interesting and I'm going to start playing with it.

I also have a similar (same?) thought as @remmycat. One thing I'd like is to be able to type a nushell table query, like

```
$mytable | where ... | select ...
```

and see the table change dynamically as I type the nushell code. (I initially thought I could hack something like this together with `fzf --preview` but I need it to be in the same nushell process because my table is large and loaded from disk).

dandavison commented:
There's no urgency to choose the final name but another possible name for this would be `interact`.

(perhaps ultimately this would go behind a "table" prefix: `$mytable | table interact` if we decide not to have less command overloading)

rgwood commented:
Super cool. This makes browsing nested tables and wide tables a lot nicer.

dandavison commented:
This is really nice @zhiburt. It looks beautiful, and it's already extremely useful as well as enjoyable to use. I can see this evolving all sorts of extra sophistication.

zhiburt commented:
> I also have a similar (same?) thought as @remmycat. One thing I'd like is to be able to type a nushell table query, like
> 
> ```
> $mytable | where ... | select ...
> ```

Played a little bit with this.

![image](https://user-images.githubusercontent.com/20165848/201497885-2f36b0a9-37a4-4656-bd8f-0e181bf44fc0.png)

> I have a very large table and so I needed a way to exit without it printing the table. I've hacked this as follows for now

Added a flag for this behavior `--peek`, so by default nothing will be returned.

fdncred commented:
wow, that's so cool, being able to type a pipeline in the tui!!!! great work!

dandavison commented:
Cool! I'm going to play with that this evening.

(Small thing: I wanted to include hyperlinks but it seems like they're stripped. Would be nice for my purposes if not too hard.)

dandavison commented:
The `:nu` query functionality is fantastic @zhiburt! I'm not just enjoying it but using it for work. OK my next feature request: Up/Down arrow to retrieve a previous command from `:nu` query history. I know that really what we want is to integrate reedline properly here, but if possible to quickly hack in Up/Down arrow history navigation to the existing implementation that would be really useful.

remmycat commented:
Wow, it's really amazing to see what you conjured up here during the last days! Thanks a lot for responding to my feedback

I love the possibility to return the last table, I didn't even think about the case that this might produce output! 😀 

I still believe it might be really valuable to not have the :nu filters be imperative commands but build up a single command chain - where you can delete parts to return to the previous views. And selecting cells interactively adds to the chain.

That way it would be kind of like a filepath or url where it's state always reflects the current view. You can either add/remove by navigating interactively, or by modifying the command directly.

There are likely problems with this I'm not thinking of, but _if_ we wanted that behaviour, it might make sense to consider it early. It would probably lead to other features working slightly differently.

e.g. the search would need to be implemented using the "find" (or a different) command, so that it can be appended to the chain and reliably produce the same output as people entering it by hand.

zhiburt commented:
> OK my next feature request: Up/Down arrow to retrieve a previous command from :nu query history. I know that really what we want is to integrate reedline properly here, but if possible to quickly hack in Up/Down arrow history navigation to the existing implementation that would be really useful.

I did integrated reedline.
But it contains now a known issue with bad rendering of last 2 lines in some scenarious, I was trying to track it down but failed so far.
And overall I am not sure whether it's good integration. (without mandatary terminal `Clear` it was not working).

So I don't know maybe a handy history buffer would be just OK.
But I left as it is......... maybe the issues will be figured out.

> (Small thing: I wanted to include hyperlinks but it seems like they're stripped. Would be nice for my purposes if not too hard.)

As far from what I've seen I assume it's not supported by `tui`? (https://github.com/fdehau/tui-rs/issues/678)

fdncred commented:
I think this is pretty close to landing as a MVP. One thing that might be helpful is that we have a status bar that has some information in it like `Press H for help` and it pops up all the shortcut keys like `arrow keys, i, /, ctrl-d`, etc.

I think we have a status bar but I'm not sure how it's used right now (the black on white bar at the bottom).

zhiburt commented:
> One thing that might be helpful is that we have a status bar that has some information in it like Press H for help and it pops up all the shortcut keys like arrow keys, i, /, ctrl-d, etc.

![image](https://user-images.githubusercontent.com/20165848/201939685-5a9bb5fa-bd96-4cc2-9501-a2f154ac7061.png)

> I think we have a status bar but I'm not sure how it's used right now (the black on white bar at the bottom).

1. Show stats when searching.
2. Shows percent of rows were seen (above a screen).

fdncred commented:
oh, that help screen looks nice! good work!

fdncred commented:
@zhiburt - I kind of feel that this isn't quite finished but I'd still like to get it into the hands of people before the next release. How are you feeling about the current state of it, are you ready to land as a MVP and continue polishing on it in later PRs?

fdncred commented:
@zhiburt no worries at all. We're super excited about this PR but we never want to force something in. We can wait until you're comfortable and have time to get things in order. As you know, things change pretty fast around here.

fdncred commented:
I've been playing with this again this morning. Still loving it. Need to work on the colors but everyone will have a different opinion on that.

I had one new idea that would be nice to add at some point. Sometimes you're browsing data and it's a bunch of columns and 1 row. It would be nice to have a keybinding like `t` to transpose that into 1 column and a bunch of rows. It would just make it easier to read instead of having to scroll right sometimes.

zhiburt commented:
> I had one new idea that would be nice to add at some point. Sometimes you're browsing data and it's a bunch of columns and 1 row. It would be nice to have a keybinding like t to transpose that into 1 column and a bunch of rows. It would just make it easier to read instead of having to scroll right sometimes.

Interesting idea.
So column then would have no name correct?

I started to believe it's in a better place now.
Surely requires some tuning here and there but overall I take it to be OK.

Added a help for commands.
And not sure if you noticed a default screen.

![download (13)](https://user-images.githubusercontent.com/20165848/203603871-6ad03816-1140-44c4-acd3-a62a26ea017d.gif)

fdncred commented:
> So column(s) then would have no name correct?

Well, it would need a column but it would be generic like `column0` or `col0`

> I started to believe it's in a MUCH better place now.

Agreed. Each time I play with it, it just makes me smile.

> Surely requires some tuning here and there but overall I take it to be OK.

Sure. That is expected. We can continue to iterate on it in other PRs.

> Added a help for commands.

Yes, I saw that and it looks great. One future change is that I'd like to to work with `:help` and `:h`, just for convenience.

> And not sure if you noticed a default screen.

Oh, no. I hadn't noticed that.

> A NOTE: Not sure if you've got chance to notice it already but we kind of grow views endlessly unless you get out from it. And I think it's not always the case sometimes you don't want to get back to the previous view EVER.

If you mean that we can drill down into as many levels are available and this just hit esc esc esc esc, etc. Then, yes, I noticed that and it seems to work great!

> Is there any like idea where this behavior would be useful and if so how should it be allied (I mean a command/shortcut or a default flow in some cases)?

Maybe I'm not understanding this.

fdncred commented:
oh, I just noticed that `:try` isn't in your help screen.

zhiburt commented:
Added support for `<t>`.

> oh, I just noticed that :try isn't in your help screen.

fixed

> Well, it would need a column but it would be generic like column0 or col0

Though I left it just as enumeration.
And I did include a column with a header names.

> Maybe I'm not understanding this.

So for example you enter scroll right from nushell typing `scroll`.
You get an a welcome screen.
You do something later.

And I tend to believe that the welcome screen is NO LONGER valuable.
Meaning we'd better not get back to it by pressing `ESC`.

What do you think?
And I am asking is there any other place where it's applicable?

fdncred commented:
> Added support for <t>.

good job! that was super quick!

> And I tend to believe that the welcome screen is NO LONGER valuable.

ya, you're probably right. It seems good to have the welcome screen if you just type `scroll` but once you start using it maybe it goes away. So, it kind of acts like a splash screen that you just see once and only if you invoke scroll directly without a pipeline.

fdncred commented:
ci is green. does that mean we're ready to land this?

zhiburt commented:
> ci is green. does that mean we're ready to land this?

I believe so.

The only thing I would note somewhere about customization which can be applied.

![image](https://user-images.githubusercontent.com/20165848/204001808-76ba180d-b859-4d47-949b-2b344dd36d74.png)

https://github.com/nushell/nushell/blob/23b8bcf23c5093007ee892b88b6758cd97d9f1f4/crates/nu-utils/src/sample_config/default_config.nu#L297-L310

fdncred commented:
Right. We definitely need some documentation on what `scroll` is and how to configure it.  Of course, that can be a separate PR. We can start with adding some things to this PR https://github.com/nushell/nushell.github.io/pull/666. Maybe a gif and some notes on what it does?

zhiburt commented on asda.txt:
```diff
@@ -1,5 +0,0 @@
-nu asdsa asdasd 
```
How this file got here .... :smile: 
Sorry

zhiburt replied:
The same must be done for `nu-scroll.txt`

fdncred commented:
@zhiburt sorry - i've broken the fmt with my small tweaks - can you please fix that?

theAkito commented:
Will this be off by default?

fdncred commented:
@theAkito this is a command that is neither on nor off by default. It's just a new command.

theAkito commented:
@fdncred

Right, I got carried away by the issue talking about it being implicit or something. I get it now.

fdncred commented:
@zhiburt while the team was testing this PR we found a small glitch and was hoping you'd be able to fix it. We'd like to be able to do `open --raw Cargo.toml | scroll`. This is really the `less` type of paging functionality. Thoughts?

zhiburt commented:
> @zhiburt while the team was testing this PR we found a small glitch and was hoping you'd be able to fix it. We'd like to be able to do open --raw Cargo.toml | scroll. This is really the less type of paging functionality. Thoughts?

So back then I was thinking about such situation.
And you can open the entry then you'll see the content split by lines.

(I've already forgot that it was done)

zhiburt commented:
Aaa or you mean this thing that we shall show only 1st line right?.

<img width="998" alt="image" src="https://user-images.githubusercontent.com/20165848/204055454-cf2c4d4a-888d-4e24-b177-70e5418bfdbd.png">

zhiburt commented:
So I fixed it by showing only 1 line. Which to be a candor might be misleading.
I would probably stick with an old approach but IDK.

<img width="156" alt="image" src="https://user-images.githubusercontent.com/20165848/204056633-98e10e7d-ec91-46cb-be9f-68bb3083583a.png">

fdncred commented:
> Aaa or you mean this thing that we sort of combined lines together.

Ya. I think if this was split by line ending as if you had typed `open --raw Cargo.toml | lines | scroll`. Then it'd probably be ok.  It's really the `cat file | less` functionality the team was talking about.

zhiburt commented:
> It's really the `cat file | less` functionality the team was talking about.

Press `<i>` enter and you'll make it so.

We could expand it by default if it's a 1x1 matrix.
BUT I kind of feel it's not that good.

fdncred commented:
Oh ya, I see the `<i><enter>` functionality does this. I'm thinking that it may need to do that automatically, again, to have the paging type of functionality without making people stumble on how it works.  It's kind of foreign to have to push i and enter to get paging to work.

There are some other future changes we may want to do that make it work a bit differently. I'm not asking for these changes right now, but just for you to think about this concept. Right now it works like a table where you can choose a column, hit enter, and drill down on it to see more. What if, instead of a table, there was maybe a different mode `scroll -t` where it was more of a tree structure. So, it kind of combines this horizontal and vertical paging with a `table -e` feel of being able to drill down in the tree. Thoughts? This was one idea that @jntrnr had. Maybe they can explain it better.

zhiburt commented:
> I'm thinking that it may need to do that automatically

Imagine there's some program.
Which does output `STDOUT` but sometimes it outputs to `STDERR` too.

If we do it automatically.
In first case there would be an output of STDOUT by lines. (like you propouse)
But in a second case we would get a 2 columns with 1 entry.

What I am saying that behavior would be different dependent on the input.
Is it OK?
I am kind of thinking being a deterministic is better here.

> where it was more of a tree structure.

Hardly can picture it.

If it suppose to be like an original `-r`
What would you say about this picture.

If we would love to get elements on the right only this table must be changed and others sort of keep it's original content correct?

<img width="835" alt="image" src="https://user-images.githubusercontent.com/20165848/204058584-d5757357-7a87-4221-ac3d-7374a326a7b0.png">

sholderbach commented on crates/nu-scroll/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+
+terminal_size = "0.2.1"
+strip-ansi-escapes = "0.1.1"
+reedline = { version = "0.13.0", features = ["bashisms", "sqlite"] }
```
This is out of date (currently `0.14`)

sholderbach commented on crates/nu-scroll/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+strip-ansi-escapes = "0.1.1"
+reedline = { version = "0.13.0", features = ["bashisms", "sqlite"] }
+crossterm = "0.24.0"
+tui = "0.19.0"
```
Sadly we are still behind the most recent version `0.25` of `crossterm` (haven't gotten around to the necessary changes in `reedline`). `tui` is already on the most recent version, so as it stands this introduces dependency duplication.

zhiburt replied:
scroll actually doesn't use reedline I believe.
But `nu-cli` does.

I guess to mitigate this issue we could downgrade tui version, right?

zhiburt replied:
Funny thing is that a previous version of `tui` uses `crossterm` `0.23`

jntrnr commented:
@zhiburt - thanks for the work on this! lots of cool stuff. As @fdncred mentioned, after he mentioned this was ready I took it for a spin and had a bunch of ideas for it. He's mentioned some, but I wanted to list them here so we could chat about them.

* working for records

This one seems pretty important, that it would work for records the same way that `table` currently works for records. You describe one idea I was also wondering about, though there may be a simpler first step on the way there. That way we can continue to evolve it. If `| scroll` on a record let you page around inside of the `table` or `table -e` output, that's already better than what the PR currently is doing, which is to put everything into one row.

You're right that we might want a way to navigate right, or explore subtrees, too. Maybe we can work towards that across multiple PRs?

* @sholderbach mentioned the version things, hopefully they're an easy fix

* I noticed some odd completion issues trying out this branch

I wasn't sure why it was happening, but I tried typing `sc<tab>` to see the completions and `scroll` didn't show up. Does it show up for you? Not sure what it is. If you can't repro it, I'll try to dig a bit deeper to figure out why it's happening.

* scrolling around large text

I think @fdncred mentioned this also above. When I was reading the title of the PR, I was wondering if it let me scroll around large strings (like text files) but when I tried to do something like `open Cargo.toml --raw | scroll` it didn't let me scroll around inside of it.

I think my mental model is that the scrolling in a way "does the right thing" for the inputs it's given. You've got the table scrolling in there (including search, which is awesome to see!). I'm betting folks will want to try it out with other types of input too, to see what it will do.

dandavison commented:
In addition to scrolling, you can drill into the table, and use queries to filter / search, and other things. What do people think the name should be? My suggestion is `interact`.

jntrnr commented:
Played with it a bit more. This thing is great. I'm betting people are really going to enjoy it.

I think there are a few things we should probably do before we land the PR, and then we can continue improving from there. Chatting a bit with @fdncred I think the main one is handling the different inputs.

For inputs we don't have an interactive view mode specifically for, I think we should run the data through `table` (or `table -e`) first, and then scroll around the results. This isn't as good as a full interactive mode, but I think it'd make a nice step on the way to having more full interactions. This would be like a fallback mode. Records and binary data were two that I saw that didn't yet get displayed in the same way as `table` does. If it's possible, I think we could take the string that `table` outputs it and then use that as the data we're scrolling around.

You can see what I mean if you do `sys | table -e | scroll`. While it lacks the colours (which would be nice to add), it does feel like you're scrolling around the table output.

Then, in future PRs, as we come up with ways of making the record explorer interactive, then we can improve it.

zhiburt commented:
Hi there,

> If | scroll on a record let you page around inside of the table or table -e output, that's already better than what the PR currently is doing, which is to put everything into one row.

> For inputs we don't have an interactive view mode specifically for, I think we should run the data through table (or table -e) first, and then scroll around the results. 

I might miss understood your point, but,
I tend to believe `less -S` will do just that?

But I've added it as a command.
Allthough I am not sure whether it brings a lot to a table? (except that it can be used to scroll of a content of big strings which is not currently possible in table view).

![download (16)](https://user-images.githubusercontent.com/20165848/204335596-50fbb297-67b7-4bee-8504-b07197c854cf.gif)

Colors are yet not possible to do as it requires a small PR to `tui-rs`.

But as it was a few times said I tend to think that these 2 concepts (`less`/`more` and interactive view via picking etc.) are a bit apposite and it might be better to be spitted as different commands.

> I wasn't sure why it was happening, but I tried typing sc<tab> to see the completions and scroll didn't show up. Does it show up for you? Not sure what it is. If you can't repro it, I'll try to dig a bit deeper to figure out why it's happening.

Works for me 

> In addition to scrolling, you can drill into the table, and use queries to filter / search, and other things. What do people think the name should be? My suggestion is interact.

Maybe `interact` is too generic?

I just like `scroll` more,
though it's not totally comply with a @webbedspace argument any more

zhiburt commented:
> We were talking earlier about some heuristic approach that could determine the type of pipelinedata coming in and automatically start in table mode or preview mode based on that data. It may not even need to be a heuristic. you may be able to just check the nushell data type and enable a mode based on that type. Thoughts?

It's a few lines of code change,
But do you certain you want to do it? (As I've still been feeling that sticking to a default is better)

> One thing that less cannot do is determine when to turn off ... because if we have horizontal scrolling, there really isn't much need to truncate columns. So, that's a plus in favor of preview mode.

You could do `table -w 9999999` and they'd never appear.

> What do you think about `explore`

`explore` is good too.

fdncred commented:
> But do you certain you want to do it? 

I'm not certain but I think JT wants to try this out first. JT usually has good intuition about these things so I'm willing to try it. If it ends up being bad, we can revert it.

> You could do table -w 9999999 and they'd never appear.

Good point. I hadn't considered that.

> explore is good too.

Let's go with that then, please.

zhiburt commented:
> I'm willing to try it.

Must be done

> Let's go with that then, please.

Renamed to `explore`

fdncred commented:
Thanks @zhiburt. I tested with `open -r Cargo.toml | explore` and it went into "preview" mode. Then I did `open Cargo.toml | explore` and it went into table mode. I think this is what JT was looking for.

One quick question, Is there a way to debug the keybindings? I can't get pageup and pagedown to work in Windows in preview mode.

zhiburt commented:
> One quick question, Is there a way to debug the keybindings? 

I don't think so but how it would look?

> I can't get pageup and pagedown to work in Windows in preview mode.

There was no such keybindings, but I've added it. 

> it's in table mode vs preview mode (had to do i enter to see the file). this should probably auto detect too if possible. pageup/pagedown work fine here.

Made it so.

fdncred commented:
> I don't think so but how it would look?

As long as the keybindings work, there's no need for it. I was just wondering if there was something like `keybindings listen` for the tui. We don't really need it though since it wasn't a conflict that caused pageup/pagedown not to work.  Thanks for adding that!

fdncred commented:
Playing with it again this morning. Found a small bug.

1. open -r cargo.toml | explore
2. pageup pagedown (yay it works!)
3. :try
4. ls | sort size <-- boom, red error
5. figured out that i should've typed sort-by. i try sort-by and it works, but the red error is still there. that is the bug.

kubouch commented on crates/nu-explore/src/commands/try.rs:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    pub const NAME: &'static str = "try";
```
Unfortunately, we just added `try` in 0.72 as a part of `try { ... } catch { ... }`. Could you pick a new name?

zhiburt replied:
It's unrelated to it.

It's just a command name, so you can type `:try`.

kubouch commented on crates/nu-cli/src/util.rs:
```diff
@@ -286,6 +286,34 @@ pub fn eval_source(
     true
 }
 
+pub fn eval_source2(
```
Seems like the intent of `eval_source2` is to strip away the printing functionality from the original `eval_source()` and the display hook and to be used to evaluate the commands inside the pager, right? I'm thinking that a) it could probably use a better name and b) shouldn't it be a part of the pager instead of changing Nushell internals?

zhiburt replied:
> Seems like the intent of eval_source2 is to strip away the printing functionality from the original eval_source()

Yes

>  a) it could probably use a better name and

Not the worth name though :)

> b) shouldn't it be a part of the pager instead of changing Nushell internals?

It'd better be, if it can be moved it should.
*I'll check it.

zhiburt replied:
moved.

kubouch commented:
We all agreed to land this, just a few small things. Could you write up what user-facing changes you introduced? Seems like there is at least the `explore` command added. Also there is a bunch of other commands available inside the `nu-explore` crate, this would be great to document as well (seems like they're not added to the default Nushell context). Some small user guide (like that you can type `:help`) that we could put into release notes would be useful as well.

Thanks for the hard work, it really feels amazing and is a huge step forward in Nushell's interface!

zhiburt commented:
> Playing with it again this morning. Found a small bug.
> 
> 1. open -r cargo.toml | explore
> 2. pageup pagedown (yay it works!)
> 3. :try
> 4. ls | sort size <-- boom, red error
> 5. figured out that i should've typed sort-by. i try sort-by and it works, but the red error is still there. that is the bug.

Fixed.

It's actually was not a bug but a logic :)
I mean when everything go well there's no point in doing so doesn't update the bar.

> 1. in :try mode, it would be great to highlight the rectangle that has the focus. it's hard to tell when you can hit i versus type i

Not sure what is the best way.
But addressed it.

> 2. mouse wheel scrolling would be nice

Not sure whether it's good but made is so.

zhiburt commented:
> We all agreed to land this, just a few small things. Could you write up what user-facing changes you introduced? Seems like there is at least the `explore` command added. Also there is a bunch of other commands available inside the `nu-explore` crate, this would be great to document as well (seems like they're not added to the default Nushell context). Some small user guide (like that you can type `:help`) that we could put into release notes would be useful as well.

Only `explore` command was added.
No other changes were supposed to be made.

Yes I guess some sort of manual would be beneficial.

kubouch commented:
> > We all agreed to land this, just a few small things. Could you write up what user-facing changes you introduced? Seems like there is at least the `explore` command added. Also there is a bunch of other commands available inside the `nu-explore` crate, this would be great to document as well (seems like they're not added to the default Nushell context). Some small user guide (like that you can type `:help`) that we could put into release notes would be useful as well.
> 
> Only `explore` command was added. No other changes were supposed to be made.
> 
> Yes I guess some sort of manual would be beneficial.

I mean, could you put these into the PR description (see the PR template)? It makes it easier for us when compiling the release notes and if somebody wants to look up the PR later for some reason.

kubouch commented:
It also introduced some new config settings, these are also user-facing changes.

fdncred commented:
I updated the OP's issue text a bit, just to change the name to `explore` and add the `config` section.

fdncred commented:
@zhiburt I found another glitch. Let me know if this is supposed to work or not. I can't do `:nu $nu`. It just gives an error.

BTW - I like how you highlighted the borders in double-lined borders. Nice work.

kubouch commented:
For me `:nu $nu` works in an empty `explore` but I'm getting a panic on `:nu`.

@zhiburt Sorry for the ping, the steps to reproduce the panic:
```
> explore
# inside explore:
:nu
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', crates/nu-explore/src/commands/nu.rs:87:41
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

fdncred commented:
oh, `:nu $nu` is working for me now too. I wonder what combination of things made it not work for me earlier.

zhiburt commented:
> LOL - I just found something else fun. Make sure you're in try mode and type clear. Not quite the behavior we want. haha

Made it being ignored.

> For me :nu $nu works in an empty explore but I'm getting a panic on :nu.

Ohhhhh sorry.

Must be fixed.

fdncred commented:
just fyi @zhiburt - i'm compiling information based on feedback from the team. i'll tag you in that issue when i create it.

qfel commented on crates/nu-explore/src/events.rs:
```diff
@@ -0,0 +1,51 @@
+
+    pub fn next(&self) -> Result<Option<KeyEvent>> {
+        let now = Instant::now();
+        match poll(self.tick_rate) {
```
Why the need to poll? This causes the UI to redraw 4 times/s (I was trying to debug something and boy did the logs spam).

I changed it to a simple blocking read and things seem to work just fine (if not smoother - but maybe that's placebo).

fdncred replied:
@rgwood may have an idea too since he's done some of the latest work on explore.

I figured this was just based on the tui example when maxim wrote it a few years ago. I'd be interested in seeing/trying your alternative.

qfel replied:
> I'd be interested in seeing/trying your alternative.

https://github.com/nushell/nushell/pull/13574
